crates/compiler/parse/src/src64.rs

```diff
@@ -0,0 +1,556 @@
+                                hFile: *mut c_void,
+                                lpBuffer: *mut c_void,
+                                nNumberOfBytesToRead: u32,
+                                lpNumberOfBytesRead: *mut u32,
+                                lpOverlapped: *mut c_void, // this should be a pointer to a struct, but we always pass null.
+                            ) -> i32;
+                        }
+
+                        let mut bytes_read = core::mem::MaybeUninit::uninit();
+
+                        // We should have already errored out if file_size exceeded u32::MAX,
+                        // due to our maximum source file size. This debug_assert! is here to
+                        // make sure casting file_size to u32 is safe in the ReadFile call.
+                        debug_assert!(MAX_ROC_SOURCE_FILE_SIZE <= u32::MAX as usize);
+
+                        ReadFile(
+                            file.as_raw_handle() as *mut c_void,
+                            buf.as_ptr() as *mut c_void,
+                            file_size as u32,
+                            bytes_read.as_mut_ptr(),
+                            core::ptr::null_mut(),
+                        );
+
+                        bytes_read.assume_init() as usize
+                    }
+                };
+
+                // We can close the file now; we're done with it.
+                drop(file);
+
+                // It's crucial that we successfully read the entire file; otherwise, it would be unsafe
+                // to make a slice out of it because we might not have overwritten the uninitialized
+                // memory leading up to the newlines at the end!
+                //
+                // Note that on UNIX, bytes_read might be -1 if this was a file read error. This
+                // condition will catch that too, since we know file_size won't be (-1isize as usize)
+                // beacuse if it was, then this match would have taken the None branch due to
+                // (-1isize as usize) exceeding our maximum file size.
+                if bytes_read != file_size {
+                    return Err(FileErr::ReadErr);
+                }
+
+                // Before we write newlines to the last chunk, branchlessly prefetch the first four 64-byte chunks.
+                // We're about to have a cache miss due to loading the last chunk from main memory (DMA will have
+                // written it there without having gone through the CPU), and if we don't prefetch here, then we'll
+                // immediately get a second cache miss when we start traversing the loaded file. The prefetch means
+                // by the time we finish resolving the first cache miss on the last chunk, continuing with the first
+                // chunk(s) won't be a cache miss anymore because they'll already be in cache.
+                //
+                // We can do further prefetches in the actual tokenization loop.
```

Comment:
Prefetching here seems like it would avoid a second cache miss (for the reason mentioned in the comment), but once I have everything working I can try removing this and see whether it actually made a difference in practice.